### DISCUSSION TITLE
valibot support openapi

### DISCUSSION THREAD

**Discussion by PandaWorker** [open-circle/valibot#149] [Ideas]
I really like valibot. I understand that it is still at the development stage.

Will there be support for generating openapi schemas in the future?
I know a very good Python library, pydantic, everything is implemented there very conveniently. From class-based validation and openapi generation out of the box.

If there is openapi support out of the box it will be the best fastify + valibot combination

**Comment by fabian-hiller** [comment]
Can you give me more context on this? What exactly is an OpenAI schema?

**Reply by PandaWorker** [reply]
Open API

https://spec.openapis.org/oas/v3.1.0
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md
https://json-schema.org/draft/2020-12/release-notes.html

https://www.npmjs.com/package/zod-openapi
https://www.npmjs.com/package/@anatine/zod-openapi

**Reply by PandaWorker** [reply]
You can see how this is implemented by pydantic
https://docs.pydantic.dev/latest/usage/json_schema/#json-schema-types

That is, we need to generate entities that will be used in ajv schema. 

Which generates the OpenAPI schema.

Here is an example https://petstore.swagger.io/

**Reply by PandaWorker** [reply]
https://fastify.dev/docs/latest/Reference/Validation-and-Serialization/

Usage example:
```
const RegistrationBodySchema = object({
    username: string().openApi({description: 'a', defult: 'anonym'}),
    password: string()
});

const opts = {
  schema: {
      body: RegistrationBodySchema.openApi
    }
  }
}

fastify.get('/', opts, (request, reply) => {
  reply.send({ params: request.body}) // echo the body
});
```

**Reply by fabian-hiller** [reply]
Without understanding everything, it seems to be out-of-scope to generate an OpenAI schema. Valibot is primarily a schema library to validate data. However, with our modular design, it seems possible to me to build this functionality on top of Valibot in a separate project.

**Reply by bingtsingw** [reply]
There is an [zod-to-openapi](https://github.com/asteasolutions/zod-to-openapi) lib maybe we can learn from it.

**Reply by fabian-hiller** [reply]
That's cool. If anyone wants to create such a package or port the code from this one to Valibot, feel free to add it to our ecosystem page: https://valibot.dev/guides/ecosystem/

**Reply by ironnysh** [reply]
There's [this tool](https://github.com/OpenAPITools/openapi-generator). Maybe it can help

**Comment by camflan** [comment]
I have started a project: https://github.com/camflan/valibot-openapi-generator

I'm using it in production for a client project but it definitely needs help to be where I'd like it to be 😅

**Reply by fabian-hiller** [reply]
Amazing! Feel free to reach out if you have any questions. Also, feel free to add your project to our ecosystem page: https://valibot.dev/guides/ecosystem/

**Reply by fabian-hiller** [reply]
Not sure if this helps, but feel free to take a look at the source code of `@valibot/to-json-schema`, which converts Valibot schemas to JSON schema: https://github.com/fabian-hiller/valibot/tree/main/packages/to-json-schema

**Comment by fabian-hiller** [comment]
There is a OpenAPI package for Hono that supports Valibot: https://github.com/rhinobase/hono-openapi

Here is a related article with more details: https://dev.to/mathuraditya7/introducing-hono-openapi-simplifying-api-documentation-for-honojs-2e0e
